### PR TITLE
Add TargetSynced condition to Zendesk source

### LINKED ISSUE #62: Zendesk: expose status of Trigger/Target as a condition
The generic source status only exposes  2 conditions: `SinkProvided` and `Deployed`.
It would be nice to have an extra one that is specific to Zendesk's dependencies (e.g. `TriggerSynced`).

### PR DESCRIPTION
**Pull request by antoineco**
Adds a `TargetSynced` condition to `ZendeskSource`.

The condition is True if everything went fine, and False in case of error.
There are 3 possible reasons for erroring during the target/trigger sync:
* `MissingAdapterURL`
* `MissingSecret`
* `FailedSync` (generic Zendesk API failure)

Demo:

```console
$ kubectl get zendesksources.sources.triggermesh.io
NAME            READY   REASON       URL   SINK                                        AGE
zendesksource   False   FailedSync         http://sockeye.default.svc.cluster.local/   4h4m
```

```console
$ kubectl describe zendesksources.sources.triggermesh.io
  Conditions:
    (...)
    Last Transition Time:  2020-07-23T17:34:21Z
    Message:               Unable to create Target
    Reason:                FailedSync
    Status:                False
    Type:                  TargetSynced
  Events:
Events:
  Type    Reason        Age   From                      Message
  ----    ------        ----  ----                      -------
  Warning InternalError 14m   zendesksource-controller  error creating Zendesk target: Zendesk API returned error (422): {"error":"RecordInvalid","description":"Record validation errors","details":{"url":[{"description":"URL: is invalid","error":"InvalidValue"}]}}
```

Closes #62

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -12,10 +12,8 @@ replace (
 
 require (
 	github.com/cloudevents/sdk-go/v2 v2.0.1-0.20200630063327-b91da81265fe
-	github.com/imdario/mergo v0.3.9 // indirect
+	github.com/google/go-cmp v0.4.0
 	github.com/kelseyhightower/envconfig v1.4.0
-	github.com/prometheus/client_golang v1.7.1 // indirect
-	github.com/prometheus/statsd_exporter v0.17.0 // indirect
```
Comment (antoineco) [inline go.mod:18]: I didn't add any dependency, but I ran `go mod tidy` which generated this diff.

File: pkg/apis/sources/v1alpha1/zendesksource_lifecycle.go [hunk 2]
```diff
@@ -76,3 +76,34 @@ func (*ZendeskSource) GetEventTypes() []string {
...
+// Reasons for status conditions
+const (
+	// ZendeskReasonFailedSync is set on a condition when some external resource can not be reconciled.
+	ZendeskReasonFailedSync = "FailedSync"
+)
+
+// zendeskSourceConditionSet is a set of conditions for ZendeskSource objects.
+var zendeskSourceConditionSet = apis.NewLivingConditionSet(
+	ZendeskConditionTargetSynced,
+)
```
Comment (antoineco) [inline pkg/apis/sources/v1alpha1/zendesksource_lifecycle.go:97]: This condition set is only for the Zendesk specific conditions. Generic conditions (SinkProvided/Deployed) are still managed by the common condition set.

File: pkg/reconciler/zendesksource/reconciler.go [hunk 3]
```diff
@@ -47,5 +47,11 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, src *v1alpha1.ZendeskSou
 		return fmt.Errorf("failed to reconcile source: %w", err)
 	}
 
-	return r.ensureZendeskTargetAndTrigger(ctx)
+	if err := r.ensureZendeskTargetAndTrigger(ctx); err != nil {
+		src.Status.MarkTargetNotSynced(fmt.Sprintf("Unable to sync Target: %s", err))
+		return err
+	}
+	src.Status.MarkTargetSynced()
```
Comment (antoineco) [inline pkg/reconciler/zendesksource/reconciler.go:54]: Like I said, the logic is dumb and always writes the same generic error reason "FailedSync". We could think about differentiating cases inside `zendesk.go` (e.g.  "MissingSecret", "RequestDenied", ...) as a future improvement.

**Comment by antoineco** [comment]
Converting to draft until a figure out a way to fix tests cleanly.

**Comment by antoineco** [comment]
I added a mechanism that can skip certain parts of the reconciliation in tests. Moving to _Ready for review_.

File: pkg/reconciler/common/skip/skip.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+func EnableSkip(ctx context.Context) context.Context {
+	return context.WithValue(ctx, skipKey{}, struct{}{})
+}
+
+// Skip returns whether the given context has skipping enabled.
+func Skip(ctx context.Context) bool {
+	if ctx.Value(skipKey{}) != nil {
+		return true
+	}
+	return false
```
Comment (odacremolbap) [inline pkg/reconciler/common/skip/skip.go:36]: linter is watching this

Comment (antoineco) [reply pkg/reconciler/common/skip/skip.go:36]: `gosimple` is a good linter!

Comment (antoineco) [reply pkg/reconciler/common/skip/skip.go]: Done in a269633

File: pkg/reconciler/common/skip/skip.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+// code execution. Mainly used to avoid variances while testing certain
+// functions.
+package skip
+
+import "context"
+
+type skipKey struct{}
+
+// EnableSkip returns a copy of a parent context with skipping enabled.
+func EnableSkip(ctx context.Context) context.Context {
```
Comment (JeffNaef-Triggermesh) [inline pkg/reconciler/common/skip/skip.go:27]: Some good learning for me here! thank you for this! I have not seen a 'skipKey' before today

File: pkg/reconciler/testing/reconcile.go [hunk 6]
```diff
@@ -262,9 +280,12 @@ func Populate(srcCpy v1alpha1.EventSource) {
 		Name:       addr.GetName(),
 	}
 
-	status := srcCpy.GetSourceStatus()
-	status.CloudEventAttributes = common.CreateCloudEventAttributes(srcCpy.AsEventSource(), srcCpy.GetEventTypes())
-	status.InitializeConditions()
+	// *reconcilerImpl.Reconcile calls this method before any reconciliation loop. Calling it here ensures that the
+	// object is initialized in the same manner, and prevents tests from wrongly reporting unexpected status updates
+	reconciler.PreProcessReconcile(context.Background(), srcCpy.(duckv1.KRShaped))
```
Comment (JeffNaef-Triggermesh) [inline pkg/reconciler/testing/reconcile.go:285]: very cool! Thank you again!

File: pkg/reconciler/common/skip/skip.go (hunk 5 above)
Comment (antoineco) [reply pkg/reconciler/common/skip/skip.go:27]: Me neither and I'm not proud of it! But in this particular case it's a super easy way to split the test suite in 2 parts (generic / Zendesk-specific)

**Comment by antoineco** [comment]
Thank you, sire Jeff. Your Zendesk source is getting better every day.

**Pull request merged**